src/test/script_tapscript_tests.cpp

```diff
@@ -0,0 +1,1546 @@
+    {
+        using std::vector<unsigned char>::vector;
+        explicit bytevector(std::string_view sv) {
+            resize(sv.size());
+            std::copy(sv.begin(), sv.end(), begin());
+        }
+
+        explicit bytevector(uint256 u) {
+            resize(u.size());
+            std::copy(u.begin(), u.end(), begin());
+        }
+
+        /**
+         * Return half-open subrange from byte vector: `[i:j)`
+         */
+        bytevector subrange(size_t i, size_t j) const {
+            assert(i <= j && j <= size());
+            bytevector r(j-i, 0);
+            std::copy(begin()+i, begin()+j, r.begin());
+            return r;
+        }
+
+        using std::vector<unsigned char>::operator[];
+        bytevector operator[](std::tuple<size_t, size_t> range) const {
+            auto [i, j] = range;
+            return subrange(i, j);
+        }
+
+        void concat(const bytevector& bv) {
+            size_t this_size{size()};
+            resize(size() + bv.size());
+            std::copy(bv.begin(), bv.end(), begin() + this_size);
+        }
+
+        std::string to_string() const {
+            return std::string(begin(), end());
+        }
+    };
+
+    bytevector operator ""_bv(const char* s, size_t len) {
+        return bytevector(std::string_view(s, len));
+    }
+
+    bool operator==(const bytevector& lhs, std::string_view rhs) {
+        return lhs.to_string() == rhs;
+    }
+
+    bool operator==(std::string_view lhs, const bytevector& rhs) {
+        return lhs == rhs.to_string();
+    }
+
+    bool operator<(const bytevector& lhs, const bytevector& rhs) {
+        return std::lexicographical_compare(lhs.begin(), lhs.end(), rhs.begin(), rhs.end());
+    }
+
+    bytevector operator||(const bytevector& lhs, const bytevector& rhs)
+    {
+        bytevector r(lhs);
+        r = lhs;
+        r.concat(rhs);
+        return r;
+    }
+}
+
+BOOST_AUTO_TEST_CASE(internal_test_validate_hex_to_bytes_and_back_helpers)
+{
+    // Hex digits have correct value
+    std::pair<char, unsigned char> hexit_map[] {
+                {'0',  0}, {'1',  1}, {'2',  2}, {'3',  3}, {'4',  4},
+                {'5',  5}, {'6',  6}, {'7',  7}, {'8',  8}, {'9',  9},
+                {'a', 10}, {'b', 11}, {'c', 12}, {'d', 13}, {'e', 14}, {'f', 15},
+                {'A', 10}, {'B', 11}, {'C', 12}, {'D', 13}, {'E', 14}, {'F', 15},
+    };
+
+    for (const auto& [c, d] : hexit_map) {
+        BOOST_TEST(from_hex(std::string(2, c)).value() == valtype(1, d*16 + d));
+    };
+    // Non-hex digits fail to convert
+    for (int d = 0; d < 255; d++) {
+        char dig = static_cast<char>(d);
+        static const auto hex_digits = "0123456789abcdefABCDEF"sv;
+        if (std::none_of(hex_digits.begin(), hex_digits.end(), [dig](char c) { return c == dig; })) {
+            BOOST_TEST((!from_hex(std::string{1, dig} + dig)));
+        }
+    }
+
+    // hex strings to byte vector
+    BOOST_TEST("0A"_hex == valtype(1, 0x0a));
```

Comment:
Wow, TIL about user-defined literals in C++11.

So these are cool, but is the "xx"_hex stuff actually used in taproot-specific tests anywhere here? I can't find any usages, which would make that facility (and these tests) unnecessary if I'm reading right.

Reply:
I'm going to move all of these helpers off into homes in `src/test/util` where they can be used to make unit tests easier to write/read/maintain.  But reduce the set of them at the same time - there's some redundancy, as you've noticed.  (In a subsequent push to this PR.)

Reply:
And that's done now.